PMacc: fix `constexpr math::Vector`

psychocoderHPC opened the pull request:
- fix CUDA 11.6 compile time evaluation issues by
  - unsing `{}` instead of `()` for type generation
  - constexpr lambda capture by reference instead of copy
- add `initializer_list` support and add coresponding test
- simplify the vector deduction guide

ikbuibui commented on include/pmacc/math/vector/Vector.hpp:
```diff
@@ -473,14 +492,9 @@ namespace pmacc
-                    uint32_t(sizeof...(T_Args))>>;
+        template<typename T_1, typename... T_Args>
+        ALPAKA_FN_HOST_ACC Vector(T_1, T_Args...)
+            ->Vector<T_1, uint32_t(sizeof...(T_Args) + 1u), ArrayStorage<T_1, uint32_t(sizeof...(T_Args) + 1u)>>;
```
Much nicer

ikbuibui commented on include/pmacc/math/vector/Vector.hpp:
```diff
@@ -97,19 +98,37 @@ namespace pmacc
             {
             }
 
+            constexpr Vector(std::initializer_list<T_Type> initList)
+                : Vector([&](uint32_t const i) constexpr { return *(initList.begin() + i); })
+            {
+            }
```
This would cause an out of memory access if the initList is smaller than the dim.
In general everywhere we assume that the generator can generate dim values, but it isnt a problem anywhere else.
One solution can be to ensure the initializer list has dim elements (`initList.size()` is constexpr).
In this case should we also allow construction from `initializer_list<T_Args>` if `std::is_convertible_v<T_Args, T_Type>`?

Another solution can be to simply remove this constructor. This would "just work" for the default `ArrayStorage` type. 
The potential problem with this is that the initalizer list construction would break if  `Vector` or `Storage` have a data member aside from the actual storage (idk if they might need one at some point because `Storage` wants to do some extra magic or something else), or if the underlying storage doesnt support aggregate initialization. (Both problems dont happen right now, but if they happen they will need workarounds) 
```suggestion
```
As an effect of removing the constructor we would allow construction of `Vector` even if the `initList` is
- smaller than `dim`, with this approach the rest of the vector will be default value initilaized for `T_Type`. 
- of a different type which is convertible to `T_Type`

psychocoderHPC replied:
Removing the constructor will result into the fact that you ca not use `Vectro<int,3>{{1,2,3}}` anymore.

To short list is `constexpr auto vec3 = Vector<int, 3>{{3, 2}};` will be `note #2699-D: attempt to access storage one position past the end of an array of 2 elements`.

To long list is currently supported because we never access invalid elements. If we would disallow it we could add a static_assert but I would like to move this to another PR. I am blocked agin by our CI :-(

> Another solution can be to simply remove this constructor ... (Both problems dont happen right now, but if they happen they will need workarounds)

This I would avoid because this is wasting a lot of time and e.g. CUDA 11.6 is very restrictive and avoids many implementations those normally should work.

ikbuibui replied:
I didnt think the compiler would catch `note #2699-D`. Cool. A static assert in a seperate PR sounds fine

ikbuibui replied:
> Removing the constructor will result into the fact that you ca not use `Vectro<int,3>{{1,2,3}}` anymore.

Is this now broken?
Only as a note, we should now be careful that the Vector and Storage classes don't define data members, else our aggregate initialization will fail.

psychocoderHPC replied:
I removed it from this PR and provide the fix for CUDA 11.6 only. The initt list will come with a seperate PR

psychocoderHPC committed:
PMacc: fix constexpr math::Vector

- fix CUDA 11.6 compile time evaluation issues by
  - unsing `{}` instead of `()` for type generation
  - constexpr lambda capture by reference instead of copy
